# Deliverable 1
### Research Question
Based on performance metrics, are some football players overvalued?

Hypothesis:
- Players with higher goal-scoring rates have higher valuations, and those who are valuations far exceeding the indication given by the correlation are overvalued.

Performance metrics available:
- Goals scored per game
- Assists per game
- Minutes played
- Matches starting eleven
- Number of matches team captain
- Clean sheets (for goalkeepers) (!!We can leave goalkeepers out, since their value evaluation metrics do not correspond to those of other player positions)
- Age 
- Position (except goalkeeper since those have no goal and little to no assists)
- Foot
- Height

Required datasets:
- players.csv (to combine player-ids to actual names, birthcountry(?), age, position, foot, height, market value, highest market value)
- game_lineups.csv (to get whether the player was starting and if he was team captain (maybe redundant))
- appearances.csv (for each player appearance we get goals, assists, minutes_played)
- clubs.csv (to combine club-id to actual club names) (needed?)

# Deliverable 2

Import data

In [1]:
# Import packages
import pandas as pd, os, sys

# Load datasets
cwd = sys.path[0]
players = pd.read_csv(cwd+'/archive/players.csv')
lineups = pd.read_csv(cwd+'/archive/game_lineups.csv')
appearances = pd.read_csv(cwd+'/archive/appearances.csv')
clubs = pd.read_csv(cwd+'/archive/clubs.csv')

We can combine "lineups" with "appearances", since those are match-based, and we will add the number-of-appearances-column to the "players" dataset.

First we clean and preprocess the separate datasets by doing the following:

- Look for contradictions in the datasets "lineups" and "appearances"

- We join datasets on `game_id` and `player_id`

In [2]:
# First, we inspect the dataframe

print(appearances.isna().sum())
print(lineups.isna().sum())

print(lineups.shape, appearances.shape) #(2191911, 10) (1578761, 13)
game_stats_na = lineups.merge(appearances, how='inner', on=['player_id', 'game_id'])
print(game_stats_na.info())

game_stats_na['player_name_y'] = game_stats_na['player_name_y'].fillna(game_stats_na['player_name_x'])
game_stats = game_stats_na
print(game_stats.info())

# Since `date` and `player_name` should also be the same for both datasets we can look for contradictions now we have resolved all null values

contradicting_dates_count = (game_stats['date_x'] != game_stats['date_y']).sum()

print(f"Number of rows with non-matching dates: {contradicting_dates_count}")

contradicting_names_count = (game_stats['player_name_x'] != game_stats['player_name_y']).sum()

print(f"Number of rows with non-matching names: {contradicting_names_count}")


# The data contains no contradictions for the dates, however it does contain 31986 contradictions for the names

contradicting_names = game_stats[game_stats['player_name_x'] != game_stats['player_name_y']]
print(f"Unique contradicting names: {contradicting_names[['player_name_x', 'player_name_y']].nunique()}")

# We have 640 unique names that are contradicting

appearance_id             0
game_id                   0
player_id                 0
player_club_id            0
player_current_club_id    0
date                      0
player_name               6
competition_id            0
yellow_cards              0
red_cards                 0
goals                     0
assists                   0
minutes_played            0
dtype: int64
game_lineups_id    0
date               0
game_id            0
player_id          0
club_id            0
player_name        0
type               0
position           3
number             0
team_captain       0
dtype: int64
(2191911, 10) (1578761, 13)
<class 'pandas.core.frame.DataFrame'>
Int64Index: 1342767 entries, 0 to 1342766
Data columns (total 21 columns):
 #   Column                  Non-Null Count    Dtype 
---  ------                  --------------    ----- 
 0   game_lineups_id         1342767 non-null  object
 1   date_x                  1342767 non-null  object
 2   game_id                 1342767 non-nu

Solve contradictions by character standardization

In [3]:
char_replacement = {
    '-': ' ',
    'ö': 'o',
    'ó': 'o',
    'ò': 'o',
    'í': 'i',
    'é': 'e',
    'ä': 'a',
    'ü': 'u',
    'ß': 'ss',
    'å': 'a',
    'ø': 'o',
    'ñ': 'n',
    'ç': 'c',
    'œ': 'oe',
    'æ': 'ae',
    'ė': 'e',
    'ż': 'z',
    'ł': 'l',
    'č': 'c',
    'ś': 's',
    'ź': 'z',
    'ñ': 'n',
    'ã': 'a',
    'į': 'i',
    'š': 's',
    'ž': 'z',
    'đ': 'd',
    'ć': 'c',
    'ț': 't',
    'ğ': 'g',
    'ş': 's',
    'î': 'i',
    'ă': 'a',
    'Ș': 'S',
    'Ț': 'T',
    'İ': 'I',
    'ı': 'i',
    'ё': 'e',
    'й': 'i',
    'ю': 'u',
    'я': 'ya',
    'ë': 'e',
    'ș': 's',
    'ţ': 't',
    'ï': 'i',

}

def replace_special_chars(text, replacements):
    for special_char, normal_char in replacements.items():
        text = text.lower().replace(special_char, normal_char)
    return text

game_stats['player_name_x'] = game_stats['player_name_x'].apply(replace_special_chars, args=(char_replacement,))
game_stats['player_name_y'] = game_stats['player_name_y'].apply(replace_special_chars, args=(char_replacement,))

contradicting_names_count = (game_stats['player_name_x'] != game_stats['player_name_y']).sum()
contradicting_names = game_stats[game_stats['player_name_x'] != game_stats['player_name_y']]
print(f"Number of rows with non-matching names: {contradicting_names_count}")
print(contradicting_names[['player_name_x', 'player_name_y']].nunique())

Number of rows with non-matching names: 8118
player_name_x    166
player_name_y    166
dtype: int64


After cleaning capitals and special characters we still have 166 contradictions left. After inspection we can conclude these contradictions are regarding the inclusion of middle names, no first name or no second name.

Since these do not directly affect our results we have chosen to make player_name_x the leading column since it is the most inclusive.

In [4]:
game_stats_cleaned = game_stats.drop(['player_name_y', 'date_y'], axis=1).rename(columns={'player_name_x':'player_name', 'date_x':'date'})

Looking for conflicts such as duplicate entries

In [5]:
duplicates = appearances.duplicated(subset=['player_id', 'game_id'], keep=False)
duplicate_entries = appearances[duplicates]

# We have 0 duplicate entries

appearance_counts = game_stats_cleaned.groupby('player_id')['player_id'].count().reset_index(name='count')
result_players = pd.merge(players, appearance_counts, on='player_id', how='left')

Since we are interested in the market value, we can remove all players for which we do not have an appearance count, because those players are not included in the metrics dataset.

In [6]:
result_players_clean = result_players.dropna(axis=0, how='any')

Saving the preprocessed data

In [7]:
os.makedirs(cwd+'/clean', exist_ok=True)
game_stats_cleaned.to_csv(cwd+'/clean/game_stats_cleaned.csv')
result_players_clean.to_csv(cwd+'/clean/result_players_clean.csv')